File: problems/polkagris/submissions/partially_accepted/quadratic_greedy.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
typedef pair<ll, ll> pii;
typedef vector<int> vi;
#define rep(i, a, b) for(int i = a; i < (b); ++i)
#define trav(a, x) for(auto& a : x)
#define all(x) begin(x), end(x)
#define rall(x) rbegin(x), rend(x)
#define sz(x) (int)(x).size()

vi pi(const string& s) {
  vi p(sz(s));
  rep(i,1,sz(s)) {
    int g = p[i-1];
    while (g && s[i] != s[g]) g = p[g-1];
    p[i] = g + (s[i] == s[g]);
  }
  return p;
}

vi match(const string& s, const string& pat) {
  vi p = pi(pat + '\0' + s), res;
  rep(i,sz(p)-sz(s),sz(p))
    if (p[i] == sz(pat)) res.push_back(i - 2 * sz(pat));
  return res;
}

int main() {
  cin.sync_with_stdio(0); cin.tie(0);
  cin.exceptions(cin.failbit);

  int N, T; cin >> N >> T;
  string initial, target; cin >> initial >> target;
  assert(sz(initial) == N);
  assert(sz(target) == T);

  // Overfold on the L side onto the R side is not a problem since foldability of the R side implies
  // there exist a shorter valid fold on the L side to apply instead.

  vector<bool> dp_rcut(N);
  vector<bool> dp_lcut(N);
  dp_rcut[0] = true;
  dp_lcut[0] = true;
  rep(lcut, 0, N) {
    for (int i = 1; ; i++) {
      if (lcut-i<0 || lcut+i >= N || initial[lcut-i] != initial[lcut+i]) break;
      if (dp_lcut[lcut-i]) {
        dp_lcut[lcut] = true;
        break;
      }
    }
  }
  rep(rcut, 0, N) {
    for (int i = 1; ; i++) {
      int r = N-1-rcut;
      if (r-i<0 || r+i >= N || initial[r-i] != initial[r+i]) break;
      if (dp_rcut[rcut-i]) {
        dp_rcut[rcut] = true;
        break;
      }
    }
  }

  rep(r, 0, 2) {
    for (int lcut : match(initial, target)) {
      int rcut = sz(initial) - sz(target) - lcut;
      if (dp_rcut[rcut] && dp_lcut[lcut]) {
        cout << "possible" << endl;
        return 0;
      }
    }
    reverse(all(target));
  }
  cout << "impossible" << endl;
}

File: problems/polkagris/submissions/partially_accepted/quadratic_greedy.py
N, T = [int(x) for x in input().split()]
initial = input()
target = input()

dp_rcut = [False]*N
dp_lcut = [False]*N

dp_rcut[0] = True
dp_lcut[0] = True

for lcut in range(N):
    i = 1
    while True:
        if lcut-i<0 or lcut+i >= N or initial[lcut-i] != initial[lcut+i]: break
        if dp_lcut[lcut-i]:
            dp_lcut[lcut] = True
            break
        i += 1
for rcut in range(N):
    i = 1
    while True:
        r = N-1-rcut
        if r-i<0 or r+i >= N or initial[r-i] != initial[r+i]: break
        if dp_rcut[rcut-i]:
            dp_rcut[rcut] = True
            break
        i += 1
for _ in range(2):
    for lcut in range(len(initial)-len(target)+1):
        if initial[lcut:lcut+len(target)] != target: continue
        rcut = len(initial) - len(target) - lcut
        if dp_rcut[rcut] and dp_lcut[lcut]:
            print("possible")
            exit()
    target = target[::-1]
print("impossible")
